keras_nlp/tokenizers/sentence_piece_tokenizer.py

```diff
@@ -0,0 +1,191 @@
+import tensorflow_text as tf_text
+
+from keras_nlp.tokenizers import tokenizer
+from keras_nlp.utils.tensor_utils import tensor_to_string_list
+
+
+class SentencePieceTokenizer(tokenizer.Tokenizer):
+    """A SentencePiece tokenizer layer.
+
+    This layer provides an implementation of SentencePiece tokenization
+    as described in the [SentencePiece paper](https://arxiv.org/abs/1808.06226)
+    and the [SentencePiece package](https://pypi.org/project/sentencepiece/).
+    The tokenization will run entirely within the Tensorflow graph, and can
+    be saved inside a `keras.Model`.
+
+    By default, the layer will output a `tf.RaggedTensor` where the last
+    dimension of the output is ragged after whitespace splitting and sub-word
+    tokenizing. If `sequence_length` is set, the layer will output a dense
+    `tf.Tensor` where all inputs have been padded or truncated to
+    `sequence_length`. The output dtype can be controlled via the `dtype`
+    argument, which should be either an integer or string type.
+
+    Args:
+        proto: Either a `string` path to a SentencePiece proto file, or a
+            `bytes` object with a serialized SentencePiece proto.
+        sequence_length: If set, the output will be converted to a dense
+            tensor and padded/trimmed so all outputs are of sequence_length.
+
+    References:
+        - [Kudo and Richardson, 2018](https://arxiv.org/abs/1808.06226)
+
+    Examples:
+
+    From bytes.
+    ```python
+    # Train a SentencePiece vocabulary.
+    bytes_io = io.BytesIO()
+    ds = tf.data.Dataset.from_tensor_slices(["the quick brown fox."])
+    sentencepiece.SentencePieceTrainer.train(
+        sentence_iterator=ds.as_numpy_iterator(),
+        model_writer=bytes_io,
+        vocab_size=20,
+    )
+
+    # Tokenize inputs
+    tokenizer = SentencePieceTokenizer(proto=bytes_io.getvalue())
+    ds = ds.map(tokenizer)
+    ```
+
+    From a file.
+    ```python
+    # Train a SentencePiece vocabulary and write it to a file.
+    ds = tf.data.Dataset.from_tensor_slices(["the quick brown fox."])
+    with open("model.spm", "wb") as model_file:
+        sentencepiece.SentencePieceTrainer.train(
+            sentence_iterator=ds.as_numpy_iterator(),
+            model_writer=model_file,
+            vocab_size=20,
+        )
+
+    # Tokenize inputs
+    tokenizer = keras_nlp.tokenizers.SentencePieceTokenizer(proto="model.spm")
+    ds = ds.map(tokenizer)
+    ```
+    """
+
+    def __init__(
+        self,
+        proto,
+        sequence_length: int = None,
+        **kwargs,
+    ) -> None:
+        # Check dtype and provide a default.
+        if "dtype" not in kwargs or kwargs["dtype"] is None:
+            kwargs["dtype"] = tf.int32
+        else:
+            dtype = tf.dtypes.as_dtype(kwargs["dtype"])
+            if not dtype.is_integer and dtype != tf.string:
+                raise ValueError(
+                    "Output dtype must be an integer type or a string. "
```

Comment:
"an integer type or a string" is not 100% clear to me, should we just list the allowable dtypes?